ml-agents/mlagents/trainers/torch/networks.py

```diff
@@ -400,29 +412,23 @@ def __init__(
         # Give the Actor only half the memories. Note we previously validate
         # that memory_size must be a multiple of 4.
         self.use_lstm = network_settings.memory is not None
-        if network_settings.memory is not None:
-            self.half_mem_size = network_settings.memory.memory_size // 2
-            new_memory_settings = attr.evolve(
-                network_settings.memory, memory_size=self.half_mem_size
-            )
-            use_network_settings = attr.evolve(
-                network_settings, memory=new_memory_settings
-            )
-        else:
-            use_network_settings = network_settings
-            self.half_mem_size = 0
         super().__init__(
             observation_shapes,
-            use_network_settings,
+            network_settings,
             act_type,
             act_size,
             conditional_sigma,
             tanh_squash,
         )
         self.stream_names = stream_names
-        self.critic = ValueNetwork(
-            stream_names, observation_shapes, use_network_settings
-        )
+        self.critic = ValueNetwork(stream_names, observation_shapes, network_settings)
+
+    @property
+    def memory_size(self) -> int:
+        if self.network_body.lstm is not None:
+            return 2 * self.network_body.lstm.memory_size
```

Comment:
I would prefer this to be something like 
```
self.network_body.lstm.memory_size + self.critic.lstm.memory_size
```
To make it a lot clearer. I think ideally, we should never have to use `2` anywhere in the code.
(This is fine as it is right now, but we should try to improve later)

Reply:
Totally agree, it also opens us up to having different memory architectures for Policy and Critic. 

I cleaned it up a bit so that the None check is taken care of in NetworkBody, and `networkbody.memory_size` is used by the sub-modules, so the None check doesn't have to be done here.